coreneuron/nrniv/nrn_acc_manager.cpp

```diff
@@ -42,371 +43,465 @@ void setup_nrnthreads_on_device(NrnThread* threads, int nthreads) {
-         * Remember that we are not allowed to use nt->_data but we have to use:
-         *      double *dtmp = nt->_data;  // now use dtmp!
-                #pragma acc enter data copyin(dtmp[0:nt->_ndata]) async(nt->stream_id)
-                #pragma acc wait(nt->stream_id)
-         */
+            for (int i = 0; i < nt->n_vecplay; i++) {
+                VecPlayContinuous* vecplay_instance = (VecPlayContinuous*)nt->_vecplay[i];
 
-        /*update d_nt._data to point to device copy */
-        acc_memcpy_to_device(&(d_nt->_data), &d__data, sizeof(double*));
+                /** just VecPlayContinuous object */
+                void* d_p = (void*)acc_copyin(vecplay_instance, sizeof(VecPlayContinuous));
+                acc_memcpy_to_device(&(d_vecplay[i]), &d_p, sizeof(void*));
+
+                VecPlayContinuous* d_vecplay_instance = (VecPlayContinuous*)d_p;
 
-        /* -- setup rhs, d, a, b, v, node_aread to point to device copy -- */
-        double* dptr;
+                /** copy y_, t_ and discon_indices_ */
+                copy_ivoc_vect_to_device(vecplay_instance->y_, d_vecplay_instance->y_);
+                copy_ivoc_vect_to_device(vecplay_instance->t_, d_vecplay_instance->t_);
+                copy_ivoc_vect_to_device(vecplay_instance->discon_indices_,
+                                         d_vecplay_instance->discon_indices_);
 
-        /* for padding, we have to recompute ne */
-        int ne = nrn_soa_padded_size(nt->end, 0);
+                /** copy PlayRecordEvent : todo: verify this */
+                PlayRecordEvent* d_e_ =
+                    (PlayRecordEvent*)acc_copyin(vecplay_instance->e_, sizeof(PlayRecordEvent));
+                acc_memcpy_to_device(&(d_e_->plr_), &d_vecplay_instance,
+                                     sizeof(VecPlayContinuous*));
+                acc_memcpy_to_device(&(d_vecplay_instance->e_), &d_e_, sizeof(PlayRecordEvent*));
 
-        dptr = d__data + 0 * ne;
-        acc_memcpy_to_device(&(d_nt->_actual_rhs), &(dptr), sizeof(double*));
+                /** copy pd_ : note that it's pointer inside ml->data and hence data itself is
+                 * already on GPU */
+                double* d_pd_ = (double*)acc_deviceptr(vecplay_instance->pd_);
+                acc_memcpy_to_device(&(d_vecplay_instance->pd_), &d_pd_, sizeof(double*));
+            }
+        }
+    }
 
-        dptr = d__data + 1 * ne;
-        acc_memcpy_to_device(&(d_nt->_actual_d), &(dptr), sizeof(double*));
+#else
+        if (nthreads <= 0) {
+            printf("\n Warning: No threads to copy on GPU! ");
+            return;
+        }
 
-        dptr = d__data + 2 * ne;
-        acc_memcpy_to_device(&(d_nt->_actual_a), &(dptr), sizeof(double*));
+        /** @todo: currently only checking nvidia gpu */
+        int num_gpus = acc_get_num_devices(acc_device_nvidia);
+        if (num_gpus == 0) {
+            printf("\n WARNING: Enabled GPU execution but couldn't find NVIDIA GPU! \n");
+        }
```

Comment:
lines 123-to-132 are common between unified or non-unified. They shouldn't be duplicated.